# scRNA Tutorial MA5112
`24/02/2020`

### Tutorial Overview
The tutorial will be split into 2 sections, much like the RNA-Seq tutorial. 
 * Generating files on Lugh
 * Performing analysis on local machine
 
### FYI
For your convenience, I have installed a base conda environment on the `mscstudent` account for you. It has `kallisto` + `bustools` installed to run the analysis on lugh and will automatically load on log in. This is evident by the prefix `base` attached to your command line username `(base) mscstudent@lugh:/data/MSc/2019`. 
I have also emptied all of your directories so you can track the outputs generated by the workflow. 

### Files for the tutorial:
For a scRNA-seq analysis, you will need to know which chemistry was used for the experiment. It is usually explicitly stated on the files themselves as either ```v2``` or ```v3```. This will require a different ```whitelist``` file depending on the analysis (```10xv2_whitelist.txt``` & ```10xv2_whitelist.txt``` respectively). A ```barcode whitelist``` is the list of all known barcode sequences that have been included in the assay kit and are available during library preparation.

You will also need a ```tx2gene``` file to map transcripts to genes during the bustools analysis. Take extreme care generating this file yourself, as you should filter the GTF file used to generate the `transcripts2gene.txt` file according to specific biotypes you are interested in for the study. Sarah Ennis has provided a `t2g.txt` file with non-coding RNA biotypes filtered from the list. 

Sarah has also provided the `transcriptome.idx` file for the analysis. All of these files are available in `/data/MSc/2019/Assets/`.

`fastq.gz` files for the analysis are located at `/data/MSc/2019/1k_pbmc_protein_v3_gex_fastqs`


### Before Running the Analysis:
Shell into the compute nodes reserved for the MSc class:
```bash
ISM

cd /yourdirectory/
```

### Kallisto Bus:
`kallisto bus` works with raw FASTQ files for single-cell RNA-Seq datasets. For each read the cell barcode and UMI information and the equivalence class resulting from pseudoalignment are stored in a BUS file `output.bus` stored in the output directory, along with `matrix.ec` and `transcripts.txt` which store information about the equivalence classes and transcript names for downstream processing. Run the following commands in your directory, and specify the path to the indexed transcriptome and fastq.gz files.

```bash
kallisto bus -i $INDEX_FILE -o bus_out/ -x 10xv3 -t 2 $FASTQ_R1 $FASTQ_R2
```

### Bustools Correct:
BUS files can be barcode error corrected with regards to a technology specific whitelist of barcodes. The `correct` command will correct all barcodes that are at Hamming distance 1 (i.e. one substitution) away from a single barcode in the whitelist. Run the `bustools correct` command, you will need to reference the `10xv3_whitelist.txt` file and the `output.bus` files generated by `kallisto bus`.

```bash
bustools correct -w $10xv3_WHITELIST -o bus_out/output_corrected.bus $OUTPUT.BUS
```

### Bustools Sort:
Raw BUS output from `kallisto bus` pseudoalignment may be unsorted. To accelerate downstream processing BUS files can be sorted using bustools sort. This will create a new BUS file where the BUS records are sorted by barcode first, UMI second, and equivalence class third.

Before running the command, make a `tmp/` directory in your directory. Run the following commands, referecing the `output_corrected.bus` file generated in the previous step.

```bash
mkdir tmp

bustools sort -T tmp/ -t 2 -o bus_out/output_sort.bus $OUTPUT_CORRECTED.BUS
```
### Bustools Text:
BUS files can be converted to a tab-separated format for easy inspection and processing using shell scripts or high level languages. We will be using a python script on the output file generated. Run the following commands, referencing the sorted bus file generated in the previous step.

```bash
bustools text -o output_sort.txt $OUTPUT_SORT.BUS
```

### Formatting output for scanpy
At this stage of the tutorial your directory should have the following structure:
```bash
--bdigby/
     |
     |--bus_out/
     |     |--matrix.ec
     |     |--output.bus
     |     |--output_corrected.bus
     |     |--output_sort.bus
     |     |--run_info.json
     |     |--transcripts.txt
     |--output_sort.txt
     |--tmp/
```

Copy and paste this [python script](https://github.com/BarryDigby/scRNA-Seq/blob/master/format_scanpy.py) into your directory as a new file, and call it `format_scanpy.py`.
Open `format_scanpy.py`, and edit line 8 to your directory:

``` python
#!/usr/bin/env python

gene_min = 200
gene_max = 10000

#setup working directory
import os
os.chdir("ADD_YOUR_PATH_HERE")  <- # change to os.chdir('/data/MSc/2019/your_username')
```
Before running the script, make a directory called `scanpy` to store the results. 
```bash
mkdir scanpy

python format_scanpy.py
```
Your output folder should contain:
```bash
--scanpy
    |--barcodes.tsv  
    |--gene_hist.png  
    |--genes.tsv  
    |--matrix.mtx
```

***

Do not worry about getting the files off of lugh. I will provide outputs from analysis of 5K_pbmc files that were too large to run during the tutorial. 
